### ISSUE TITLE
KSP2 Performance Issues

### ISSUE THREAD

**Issue by ZacSweers** [google/ksp#2282]
(Filing at the request of the JB folks for referencing)

I re-profiled our builds with KSP2 with 2.0.21-1.0.27. Unfortunately, across the board in a bunch of benchmarks I found that
- abi change - KSP2 is ~40-50% slower
- noabi change - KSP2 is 80-100% slower
- clean build - negligible/amortized by the clean build
- 
If I just benchmark expensive our two most expensive KSP tasks with `, KSP2 is around 20-30% slower.

Other notes:
- We give the kotlin and gradle daemons the same memory, and these benchmarks were run on overpowered AWS boxes
- None of these builds were experiencing memory pressure with the settings they had (somewhere in the dozens of GB for xmx)

time-to-task-start is more or less the same, assumes a hot configuration cache

Here's a few samples. Looking again there is some GC time creeping into some of these that I'll look closer into, but not enough to majorly skew the larger trend

![Image](https://github.com/user-attachments/assets/00fad8eb-60a8-4424-9a2b-02efb1265cca)
![Image](https://github.com/user-attachments/assets/14f4189c-17c6-43c4-b51d-dfdf05d8dec1)
![Image](https://github.com/user-attachments/assets/a70cfc8f-5649-456a-ac3a-fa243a633ce0)
![Image](https://github.com/user-attachments/assets/fc88ff88-1ead-4219-8fb6-aafb58ff44b5)

**Comment by bcorso** [comment]
> I re-profiled our builds with KSP2 with 2.0.21-1.0.27.

@ZacSweers sorry if I missed it, but what is the comparison against?

Is this comparing two different versions of KSP2 or KSP1 vs KSP2 (or something else)?

**Comment by ZacSweers** [comment]
Comparing against KSP1 yeah

**Comment by ZacSweers** [comment]
Tested again with 2.1.10-1.0.30 but still seeing significant regressions. 

ABI - ~25%-28% slower
non-abi - ~ 40%-50% slower

**Comment by ting-yuan** [comment]
Hello @ZacSweers, [2.1.20-1.0.32](https://github.com/google/ksp/releases/tag/2.1.20-1.0.32) includes a few performance fixes on the startup time of KSP2. I hope that will help overall performance and especially on non-ABI changes. Please let me know if you're still seeing any kind of performance regressions. I'd really appreciate any data / information that may guide the direction of optimizations.

**Comment by ZacSweers** [comment]
Tested again with 2.1.20-1.0.32 with mixed results.

**TL;DR**
- Smaller projects - ~20% faster
- More intensive projects (large classpaths) - still ~15-18% slower

**Comment by ansman** [comment]
I tried KSP 2 on our large project, and this is the result:

### Longest KSP task
|  | Duration |
| --- | --- | 
| KSP1 | 13.740s |
| KSP2 (1.0.31) | 25.218s |
| KSP2 (1.0.32) | 20.842s |

### Total time spend in KSP
|  | Duration |
| --- | --- | 
| KSP1 | 14m 24.981s |
| KSP2 (1.0.31) | 15m 40.957s |
| KSP2 (1.0.32) | 11m 36.319s |

### Total build time
|  | Duration |
| --- | --- | 
| KSP1 | 5m 57.983s |
| KSP2 (1.0.31) | 5m 44.658s |
| KSP2 (1.0.32) | 4m 59.995s |

So although KSP still has room for improvement, it seems that KSP2 is faster overall now.
